### PR TITLE
Rest field tests

### PR DESCRIPTION
**Pull request by IvanGrginovInf**
Added Tests:
- [x] FieldCli
- [x] FieldExample
- [x] RouteCli
- [x] RouteExample

### CODE REVIEW

File: tests/Datasets/Arguments.php [hunk 1]
```diff
@@ -23,3 +23,17 @@
...
 	[['key' => 'bold--all']]
 ]);
+
+dataset('correctFieldNameArguments', [
+	[['field_name' => 'NekiField', 'object_type' => 'post']],
+	[['field_name' => 'neki field', 'object_type' => 'post']],
+	[['field_name' => 'neki-field', 'object_type' => 'post']],
+	[['field_name' => 'neki_field', 'object_type' => 'post']],
+	[['field_name' => 'Neki Field', 'object_type' => 'page']],
+	[['field_name' => 'Neki Field', 'object_type' => 'custom-post-type']],
```
Comment (dingo-d) [inline tests/Datasets/Arguments.php:33]: English please 😄

File: tests/Datasets/Arguments.php [hunk 2]
```diff
@@ -23,3 +23,17 @@
...
+]);
+
+dataset('errorFieldNameArguments', [
+	[['field_name' => 'NekiField']],
```
Comment (dingo-d) [inline tests/Datasets/Arguments.php:37]: ⬆️

File: tests/Helpers.php [hunk 3]
```diff
@@ -44,6 +44,9 @@ function setupMocks() {
 
 	// Mock rest response handler.
 	Functions\when('rest_ensure_response')->returnArg();
+
+	// Require all WP Mock objects.
+	require_once dirname( __FILE__ ) . '/WpMocks/WP_REST_Server.php';
```
Comment (dingo-d) [inline tests/Helpers.php:49]: This file just includes the empty `WP_REST_Server` class. Why not just instantiating it here? You don't need to concern yourself with WordPress functionality in unit tests 😉

File: tests/Rest/Fields/FieldCliTest.php [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+	$field([], $field->getDevelopArgs([]));
+
+	// Check the output dir if the generated method is correctly generated.
+	$generatedField = file_get_contents(dirname(__FILE__, 4) . '/cliOutput/src/Rest/Fields/TitleField.php');
+
+	$this->assertStringContainsString('class TitleField extends AbstractField implements CallableFieldInterface', $generatedField);
+	$this->assertStringContainsString('return \'title\';', $generatedField);
+	$this->assertStringContainsString('return \'post\';', $generatedField);
+	$this->assertStringContainsString('get_callback', $generatedField);
+	$this->assertStringContainsString('rest_ensure_response', $generatedField);
```
Comment (dingo-d) [inline tests/Rest/Fields/FieldCliTest.php:47]: Add one assertion for the 'unhappy path', like that some string is not present in the copied class. In the case the string is copied you'll know you have some weird issues 😄

File: tests/Rest/Fields/FieldCliTest.php [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+});
+
+test('REST field CLI command will correctly copy the field class with arguments', function ($fieldNameArguments) {
+	$field = $this->field;
+	$field([], $fieldNameArguments);
+	$full_field_name = "{$this->field->getFileName($fieldNameArguments['field_name'])}Field";
```
Comment (dingo-d) [inline tests/Rest/Fields/FieldCliTest.php:53]:
```suggestion
	$fullFieldName = "{$this->field->getFileName($fieldNameArguments['field_name'])}Field";
```

CamelCase

File: tests/Rest/Fields/FieldCliTest.php [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+	$object_type = $fieldNameArguments['object_type'];
```
Comment (dingo-d) [inline tests/Rest/Fields/FieldCliTest.php:54]:
```suggestion
	$objectType = $fieldNameArguments['object_type'];
```

File: tests/Rest/Fields/FieldCliTest.php [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+	// Check the output dir if the generated method is correctly generated.
+	$generatedField = file_get_contents(dirname(__FILE__, 4) . "/cliOutput/src/Rest/Fields/{$full_field_name}.php");
```
Comment (dingo-d) [inline tests/Rest/Fields/FieldCliTest.php:57]:
```suggestion
	$generatedField = file_get_contents(dirname(__FILE__, 4) . "/cliOutput/src/Rest/Fields/{$fullFieldName}.php");
```

File: tests/Rest/Fields/FieldCliTest.php [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+	$this->assertStringContainsString("class {$full_field_name} extends AbstractField implements CallableFieldInterface", $generatedField);
```
Comment (dingo-d) [inline tests/Rest/Fields/FieldCliTest.php:59]:
```suggestion
	$this->assertStringContainsString("class {$fullFieldName} extends AbstractField implements CallableFieldInterface", $generatedField);
```

File: tests/Rest/Fields/FieldCliTest.php [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+	$this->assertStringContainsString("return '{$object_type}';", $generatedField);
```
Comment (dingo-d) [inline tests/Rest/Fields/FieldCliTest.php:60]:
```suggestion
	$this->assertStringContainsString("return '{$objectType}';", $generatedField);
```

File: tests/Rest/Fields/FieldCliTest.php [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+	$this->assertStringNotContainsString('example-post-type', $generatedField);
+	$this->assertStringNotContainsString('example-field', $generatedField);
+})->with('correctFieldNameArguments');
+
+test('REST field CLI command will throw error on missing / invalid arguments', function ($fieldNameArguments) {
+	$field = $this->field;
+	$field([], $fieldNameArguments);
+})->with('errorFieldNameArguments')->throws('exception');
```
Comment (dingo-d) [inline tests/Rest/Fields/FieldCliTest.php:68]: Shouldn't the exception be `Exception::class`, that is the correct exception thrown?

File: tests/Rest/Fields/FieldExampleTest.php [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+test('Field registers the callback properly', function () {
+	$action = 'field_registered';
+	Functions\when('register_rest_field')->justReturn(putenv("SIDEAFFECT={$action}"));
+
+	$this->field->fieldRegisterCallback(new \WP_REST_Server(), 'attr', new class{}, 'post');
+
+	$this->assertEquals(getenv('SIDEAFFECT'), $action);
+
+	// Cleanup.
+	putenv('SIDEAFFECT=');
```
Comment (dingo-d) [inline tests/Rest/Fields/FieldExampleTest.php:44]: 👍🏼

**Review by dingo-d — CHANGES_REQUESTED** [review]
Left some comments. Great job 👍🏼

File: tests/Helpers.php (hunk 3 above)
Comment (IvanGrginovInf) [reply tests/Helpers.php:49]: It's used in `tests/Rest/Fields/FieldExampleTest.php` like so:

```
	$this->field->fieldRegisterCallback(new \WP_REST_Server(), 'attr', new class{}, 'post');
```

That's how I used to do it in Codeception tho I guess here I could just mock the object and instantiate it here (not sure if that's possible tho I'd imagine it is).

File: tests/Rest/Fields/FieldCliTest.php (hunk 10 above)
Comment (IvanGrginovInf) [reply tests/Rest/Fields/FieldCliTest.php:68]: Pest does that already. For example this works (and only catches InvalidArgumentException):

`->throws('InvalidArgumentException');`

But I agree, having it as `Exception::class` is much better, will change!

File: .gitignore [hunk 12]
```diff
@@ -1,9 +1,9 @@
 vendor
 node_modules
 /docs
-tests/coverage/*
 
 composer.lock
 package-lock.json
 cliOutput
-.phpunit.result.cache
```
Comment (dingo-d) [inline .gitignore:9]: Why are these two removed?

Comment (IvanGrginovInf) [reply .gitignore:9]: @dingo-d Hmm didn't commit that but shouldn't this be gitignored?

Comment (dingo-d) [reply .gitignore:9]: Yes, it should be, but I'm not sure why it's marked as deleted 👀

Comment (IvanGrginovInf) [reply .gitignore:9]: Ah, now I see. Yeah not sure what happened there, re-added it to gitignore.

**Review by dingo-d — CHANGES_REQUESTED** [review]
`tests/coverage/*` should be ignored as well in the `.gitignore`

File: tests/Helpers.php (hunk 3 above)
Comment (dingo-d) [reply tests/Helpers.php:49]: Yeah, mock it on the fly in the test.

**Review by dingo-d — APPROVED** [review]

**Pull request merged**